addons/mail/models/mail_followers.py

```diff
@@ -42,9 +42,17 @@ class Followers(models.Model):
     @api.multi
     def _invalidate_documents(self):
         """ Invalidate the cache of the documents followed by ``self``. """
-        for record in self:
-            if record.res_id:
-                self.env[record.res_model].invalidate_cache(ids=[record.res_id])
+        to_invalidate = {}
+        # use read for performance:
+        # only prefetch what is needed and avoid calling fields __get__
+        for record in self.read(['res_id', 'res_model']):
```

Comment:
Instead of reading, we can use the create vals if given in the case of creation. 
Something like:
```python3
def _invalidate_documents(self, vals_list=None):
    for record in (vals_list or self.read(['res_id', 'res_model']):

def create(self, vals_list):
   res.invalidate_documents(vals_list)
```

we could also continue to keep profit of cache if res_model or res_id in cache:
```python3
for record in (vals_list or [{'res_id': rec.res_id, 'res_model': rec.res_model} for rec in self]):
```
The side effect is that more info will be prefeched in case of CacheMiss (partner_id, channel_id) but we can avoid a query in some cases. I think it's interresting to do so since it is a very small model but tell me what you think.

Reply:
The first suggestion is excellent, by using `vals_list` from `create` and not doing a query we can reduce the time from 9ms to 0.5ms for 1000 records!

The only case where this would fail is when the result of `create` is actually different than the given values. Something similar happened recently on another case (see #32460) but I don't think it is a problem here because I don't see how the value could differ in this case:

- there is no parent to Followers for a potential override, just directly Model
- both fields we are interested in are simple, non-relational, fields: Char and Integer
- they are not computed, inversed, or anything like that
- they don't have a default value (which could otherwise be taken into account anyway)


For the second suggestion I just tried but using `read` is more than 2 times faster than full prefetch and fields `__get__` on 1000 records (22ms vs 9ms).

Also just after doing `create` I don't think the value will ever be in cache so there is no point to even try to see if it is.

And since the created records are not used afterwards, there is no benefit to prefetch everything.

Reply:
In case of create we will always have the values and it is the main case we need to optimize. 

We dont really care about write performance I think.

For unlink it can be interresting since followers are unlink when a mail.thread is unlink. But in this precise case we dont need to invalidate cache since records are unlink anyway. There is maybe something smart to do in this case but I don't have a clean solution right now exept a context key but I'm not sure it is usefull to complexify invalidate more than it is.

Reply:
I don't see what else to improve, `create` was the main issue, and `write` and `unlink` will also benefit from doing it in batch.

Let me know if you see anything else, or feel free to r+ if everything seems good.